### PR TITLE
use simple ADT to encode attribute values

### PR DESCRIPTION
**Pull request by buntec**
A basis for discussion.

### CODE REVIEW

File: snabbdom/src/main/scala/snabbdom/AttrValue.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+object AttrValue {
+
+  def apply(s: String): AttrValue = StringAttrValue(s)
+  def apply(b: Boolean): AttrValue = BooleanAttrValue(b)
+
+  case class StringAttrValue(value: String) extends AttrValue
+  case class BooleanAttrValue(value: Boolean) extends AttrValue
+
+  implicit def stringToAttrValue(value: String): AttrValue =
```
Comment (buntec) [inline snabbdom/src/main/scala/snabbdom/AttrValue.scala:51]: I know this is controversial but still seems to be quite common for such cases, see e.g., [here](https://github.com/tpolecat/natchez/blob/series/0.1/modules/core/shared/src/main/scala/TraceValue.scala)

Comment (armanbilge) [reply snabbdom/src/main/scala/snabbdom/AttrValue.scala:51]: Yeah. I think it goes to @cornerman's question in https://github.com/buntec/scala-js-snabbdom/issues/14#issuecomment-1120278952

> Do we consider this library directly user facing or is it more like a low level library powering, e.g., ff4s and outwatch.

If it's a low-level library, we probably wouldn't need this, right?

**Comment by cornerman** [comment]
In general this looks good, but i am a bit concerned about allocations - introducing a wrapper for each attribute value. Does it have advantages over `Boolean | String` apart from userfriendliness? :)

**Comment by buntec** [comment]
I hear you. I would prefer, however, not to use `js.|`, just like I have been avoiding `js.Array`, `js.UndefOr` etc. My impression is, perhaps incorrectly, that they should be reserved for interop scenarios. They also take away from the pure Scala feeling in my opinion, which is very subjective of course. 

In any case, it would be great to have a set of benchmarks to be able to measure these things so that we can make an informed decision.

**Comment by cornerman** [comment]
I understand and totally agree that we need benchmarks for a good decision :+1:

For me, I am actually using quite a lot of native types (like js array), because they should be quite optimized in today browsers - and served me well when measuring performance in a huge webapp developed with outwatch.

**Comment by armanbilge** [comment]
Another strategy here, for keeping a friendly user API while avoiding allocations could be to use an encoding like Akka's `OptionalVal`.
https://github.com/akka/akka/blob/30749205a214e6e38fafe2636ea685b4c775fff2/akka-actor/src/main/scala/akka/util/OptionVal.scala

Internally, we would have:
```scala
class AttrValue(val v: String | Boolean) extends AnyVal
```

and then expose the ADT members as `unapply` methods.

**Pull request closed without merging**